cairo/ethereum/cancun/vm/precompiled_contracts/alt_bn128.cairo

```diff
@@ -49,20 +59,96 @@ func alt_bn128_add{
-    tempvar y0_value = y0_value;
-    tempvar x1_value = x1_value;
-    tempvar y1_value = y1_value;
+    tempvar ALT_BN128_PRIME = U256(new U256Struct(alt_bn128.P_LOW_128, alt_bn128.P_HIGH_128));
+    tempvar oog_err = new EthereumException(OutOfGasError);
+    let is_x0_out_of_range = U256_le(ALT_BN128_PRIME, x0_value);
+    if (is_x0_out_of_range.value != 0) {
+        return oog_err;
+    }
+    let is_y0_out_of_range = U256_le(ALT_BN128_PRIME, y0_value);
+    if (is_y0_out_of_range.value != 0) {
+        return oog_err;
+    }
+    let is_x1_out_of_range = U256_le(ALT_BN128_PRIME, x1_value);
+    if (is_x1_out_of_range.value != 0) {
+        return oog_err;
+    }
+    let is_y1_out_of_range = U256_le(ALT_BN128_PRIME, y1_value);
+    if (is_y1_out_of_range.value != 0) {
+        return oog_err;
+    }
 
-    tempvar data = data;  // Required for rust hint
-    tempvar error: EthereumException*;
-    tempvar output: Bytes;
+    // Checks that point are on the curve
+    let x0_uint384 = uint256_to_uint384([x0_value.value]);
+    let y0_uint384 = uint256_to_uint384([y0_value.value]);
+    let x1_uint384 = uint256_to_uint384([x1_value.value]);
+    let y1_uint384 = uint256_to_uint384([y1_value.value]);
+    let p0 = G1Point(x=x0_uint384, y=y0_uint384);
+    let p1 = G1Point(x=x1_uint384, y=y1_uint384);
+    tempvar a = new UInt384(alt_bn128.A0, alt_bn128.A1, alt_bn128.A2, alt_bn128.A3);
+    tempvar b = new UInt384(alt_bn128.B0, alt_bn128.B1, alt_bn128.B2, alt_bn128.B3);
+    tempvar modulus = new UInt384(alt_bn128.P0, alt_bn128.P1, alt_bn128.P2, alt_bn128.P3);
+    tempvar g = new UInt384(alt_bn128.G0, alt_bn128.G1, alt_bn128.G2, alt_bn128.G3);
 
-    %{ alt_bn128_add_hint %}
-    if (cast(error, felt) != 0) {
-        return error;
+    // Checks verifying the points are on the curve.
+    let point_inf = G1Point(x=UInt384(0, 0, 0, 0), y=UInt384(0, 0, 0, 0));
+    let is_p0_zero = G1Point__eq__(p0, point_inf);
+    let is_p1_zero = G1Point__eq__(p1, point_inf);
+
+    let pair_is_zero = is_p0_zero.value * is_p1_zero.value;
+    if (pair_is_zero != 0) {
+        let (buffer: felt*) = alloc();
+        memset(buffer, 0, 32);
+        tempvar output = Bytes(new BytesStruct(data=buffer, len=32));
+        EvmImpl.set_output(output);
+        tempvar ok = cast(0, EthereumException*);
+        return ok;
+    }
+
+    tempvar is_on_curve;
+    tempvar point = p0;
+    %{ is_point_on_curve %}
+    tempvar is_p0_on_curve_uint384 = UInt384(is_on_curve, 0, 0, 0);
+    if (is_on_curve == 0 and is_p0_zero.value == 0) {
+        assert_not_on_curve(new p0.x, new p0.y, a, b, modulus);
+        tempvar err = new EthereumException(OutOfGasError);
+        return err;
     }
 
+    if (is_p0_zero.value == 0) {
+        assert_x_is_on_curve(new p0.x, new p0.y, a, b, g, new is_p0_on_curve_uint384, modulus);
+        tempvar range_check96_ptr = range_check96_ptr;
+        tempvar add_mod_ptr = add_mod_ptr;
+        tempvar mul_mod_ptr = mul_mod_ptr;
```

Comment:
apparently not requited (not written in the p1 if/else)

Reply:
because they're implicit args so they're in already well positioned, no?

Reply:
not sure about why somethime it works and sometimes not